### ISSUE TITLE
questions

### ISSUE THREAD

**Issue by bertsky** [MehmedGIT/wf_server_nf_script#1]
It's probably too early, but to verify I get this right and we are on the same page about the target:

1. In `post_processing_job`, why don't you query the job ID? https://github.com/MehmedGIT/wf_server_nf_script/blob/9e231d87b557f42aab8e1d0ad0af0776625afbc1/wf_server_example.nf#L74-L79
2. In `configure_and_consume_polling`, why don't you discern by job ID? https://github.com/MehmedGIT/wf_server_nf_script/blob/9e231d87b557f42aab8e1d0ad0af0776625afbc1/wf_server_example.nf#L120-L125
3. In `configure_and_consume_polling`, why poll https://github.com/MehmedGIT/wf_server_nf_script/blob/9e231d87b557f42aab8e1d0ad0af0776625afbc1/wf_server_example.nf#L134 instead of using a blocking recipe (as described in the Python pika on the [Web API](https://github.com/OCR-D/spec/blob/b6701659eae28c6fd2f08855714de8fe22b114a2/web_api.md) spec)?
4. Why have status as a second output and predicate the next consumer on it https://github.com/MehmedGIT/wf_server_nf_script/blob/9e231d87b557f42aab8e1d0ad0af0776625afbc1/wf_server_example.nf#L174-L175 instead of having the `exec` block return with non-zero (which should be possible in Groovy) and defer to the normal Nextflow error handling?

**Comment by MehmedGIT** [comment]
> It's probably too early, but to verify I get this right and we are on the same page about the target:

Yes, that's just a proof-of-concept and for the sake of providing a working example which can then be improved.

> 1. In post_processing_job, why don't you query the job ID?
> 2. In configure_and_consume_polling, why don't you discern by job ID?

Good point. This is how it should be implemented when there will be more than one workflow being executed and the result queue of a processor will have more than one message for different workflow jobs:

```nextflow
  def job_id = post_processing_job(...)
  def job_status = configure_and_consume_polling(result_queue, job_id)
```
Then the RabbitMQ consumer must verify that the message consumed from the result queue matches with the expected job_id. 

> 3. In configure_and_consume_polling, why poll instead of using a blocking recipe (as described in the Python pika on the [Web API](https://github.com/OCR-D/spec/blob/b6701659eae28c6fd2f08855714de8fe22b114a2/web_api.md) spec)?

Because I was not able to make it work with a blocking recipe (the way we use in ocrd-network, WebAPI, and Operandi) inside the Nextflow script withing the DSL2 standard no matter what I tried. I tried with both the `org.springframework.amqp` and `com.rabbitmq`. Check this commit which is one of the last versions and tries I had: 8e7a0189b460b2523beba9bdd1079c7c5ea47ea0

The process block was exiting before the consumer received any message. Then the consumer was stuck, so was the next process block because the job status was never updated properly. Any waiting strategies did also not help. Doing hacky stuff inside the `onMessage` method was always leading to a race condition, also described [here](https://rabbitmq-users.narkive.com/sUtYAIuY/cancel-consumers-based-on-message) - unless, as suggested, a separate Thread is created to cancel the consumer. However, the DSL2 standard of Nextflow limits the domain of the usable properties of Groovy and Java. E.g., implementing a class is not possible. Thus, there was no way to change the value of a passed parameter to a method since Java/Groovy is only [pass-by-value](https://stackoverflow.com/questions/40480/is-java-pass-by-reference-or-pass-by-value). That's the reason why I also tried to utilize a global variable as a last try although it does not make much sense.

If you have any ideas how the 8e7a0189b460b2523beba9bdd1079c7c5ea47ea0 could be improved to work properly I can try them. 

> instead of having the exec block return with non-zero (which should be possible in Groovy) and defer to the normal Nextflow error handling?

I am not sure I understand what you mean here.

**Comment by bertsky** [comment]
Ouch, that's sounds really painful to go through and find out!

> If you have any ideas how the [8e7a018](https://github.com/MehmedGIT/wf_server_nf_script/commit/8e7a0189b460b2523beba9bdd1079c7c5ea47ea0) could be improved to work properly I can try them.

I'll give it a look, but regarding the restriction of the Groovy scripts to not define classes, couldn't you define them in external source files and [import these from the exec block](https://antunderwood.gitlab.io/bioinformant-blog/posts/building_a_dsl2_pipeline_in_nextflow/)?




> > 4\. Why have status as a second output and predicate the next consumer on it
> > instead of having the exec block return with non-zero (which should be possible in Groovy) and defer to the normal Nextflow error handling?
> 
> I am not sure I understand what you mean here.

You artificially create a status output which you can predicate on in the follow-up (consuming) processor via `when`, but why not simply have failing processors fail in their exec block so Nextflow knows immediately (at the producer)?

**Comment by bertsky** [comment]
> [8e7a018](https://github.com/MehmedGIT/wf_server_nf_script/commit/8e7a0189b460b2523beba9bdd1079c7c5ea47ea0)
> 
> The process block was exiting before the consumer received any message.

A few wild guesses:
- [default consumer timeout](https://docs.spring.io/spring-amqp/api/org/springframework/amqp/rabbit/listener/SimpleMessageListenerContainer.html#setReceiveTimeout(long)) is just 1sec, times [consecutive idle trigger](https://docs.spring.io/spring-amqp/api/org/springframework/amqp/rabbit/listener/SimpleMessageListenerContainer.html#setConsecutiveIdleTrigger(int))
- according to [this](https://github.com/spring-projects/spring-amqp/issues/1975), the blocking consumer is on a different thread; as an (even simpler) alternative, [DirectMessageListenerContainer](https://docs.spring.io/spring-amqp/api/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.html) has been created
